browser/importer/extensions_import_helpers.cc

```diff
@@ -5,92 +5,330 @@
+        true);
+  });
+
+  return true;
+}
+
+bool ImportExtensionSettings(const std::string& extension_id,
+                             const base::FilePath& source_profile,
+                             const base::FilePath& target_profile) {
+  if (!ImportLocalExtensionSettings(extension_id, source_profile,
+                                    target_profile)) {
+    return false;
+  }
+  if (!ImportIndexedDBExtensionSettings(extension_id, source_profile,
+                                        target_profile)) {
+    return false;
+  }
+  return true;
+}
+
+ImportingExtension::ImportingExtension() = default;
+ImportingExtension::ImportingExtension(ImportingExtension&&) = default;
+ImportingExtension& ImportingExtension::operator=(ImportingExtension&&) =
+    default;
+
+ImportingExtension::~ImportingExtension() {
+  if (installer) {
+    installer->AbortInstall();
+  }
+}
+
+ExtensionsImporter::ExtensionsImporter(const base::FilePath& source_profile,
+                                       Profile* target_profile)
+    : source_profile_(source_profile), target_profile_(target_profile) {
+  CHECK(target_profile_);
+  CHECK_NE(source_profile_, target_profile_->GetPath());
+}
+
+ExtensionsImporter::~ExtensionsImporter() = default;
+
+// static
+base::RepeatingCallback<ExtensionImportStatus(const std::string& extension_id)>&
+ExtensionsImporter::GetExtensionInstallerForTesting() {
+  static base::NoDestructor<base::RepeatingCallback<ExtensionImportStatus(
+      const std::string& extension_id)>>
+      g_intsaller;
+  return *g_intsaller;
+}
+
+void ExtensionsImporter::Prepare(OnReady on_ready) {
+  CHECK(extensions_.empty());
+  extensions::GetExtensionFileTaskRunner()->PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce(&GetExtensionsList, source_profile_,
+                     target_profile_->GetPath()),
+      base::BindOnce(&ExtensionsImporter::OnGetExtensionsForImport,
+                     weak_factory_.GetWeakPtr(), std::move(on_ready)));
+}
+
+bool ExtensionsImporter::Import(OnExtensionImported on_extension) {
+  CHECK(!IsImportInProgress());
+
+  for (auto& extension : extensions_) {
+    if (extension.is_installed) {
+      on_extension.Run(extension.id, ExtensionImportStatus::kOk);
       continue;
     }
-    storages[id] = extensions::value_store_util::CreateSettingsStore(
-        extensions::settings_namespace::LOCAL,
-        extensions::value_store_util::ModelType::EXTENSION, id, factory);
+
+    auto wrapper = base::BindOnce(
+        [](base::WeakPtr<ExtensionsImporter> self, OnExtensionImported callback,
+           const std::string& id, ExtensionImportStatus status) {
+          if (self) {
+            --self->in_progress_count_;
+            callback.Run(id, status);
+          }
+        },
+        weak_factory_.GetWeakPtr(), on_extension);
+
+    ++in_progress_count_;
+    if (auto installer = GetExtensionInstallerForTesting()) {
```

Comment:
do you need to copy the callback each time? maybe `auto& installer`? or move this getter out of the for-loop.